satpy/readers/generic_image.py

```diff
@@ -117,4 +117,9 @@ def mask_image_data(data):
                                 for i in range(data.shape[0])])
         data.data = masked_data
         data = data.sel(bands=BANDS[data.bands.size - 1])
+    elif hasattr(data, 'nodatavals') and data.nodatavals:
+        data = data.astype(np.float64)
```

Comment:
Do you have to convert to float here? If we can keep the initial data type and use a fill value dependent on that data type then it may work for more situations. What if you copied the nodatavals to `.attrs['_FillValue']` for integer types, but replaced them with `np.nan` for floats? This would match what we do in other readers for "category" products.

Reply:
I'm unsure what's better option. Unfortunately, NaN is only possible with floats and I have the feeling that most of the compositors only work with NaNs as masks. It seems that only ColormapCompositor cares about "_FillValue".  That's why I've done the float conversation as it is already the case if alpha channels are present.
Maybe I should introduce a new reader parameter and let the user decide how to handle nodatavals?

Reply:
I just stumbled across the following paragraph in the satpy docs (https://satpy.readthedocs.io/en/latest/dev_guide/custom_reader.html)

> Single precision floats (np.float32) is a good compromise, as it has 23 significant bits (mantissa) and can thus represent 16 bit integers exactly, as well as keeping the memory footprint half of a double precision float.
> One commonly used method in readers is xarray.DataArray.where() (to mask invalid data) which can be coercing the data to np.float64. To ensure for example that integer data is coerced to np.float32 when xarray.DataArray.where() is used, you can do ..
So converting to floats to mask data with NaN seems a common practice. As noted in the docs, float32 would be better. Conversion to floats is already done few lines above when alpha channels are handled.

Reply:
I think that is more specifically talking about integer data that is in the file as integers and needs to be scaled (ex. `data * scale_factor + add_offset`) to floats, but also has a fill value. In that case you'd need to do the `.where` call to mask out the floats with some other value. In Satpy that is NaN for floating point data and any `_FillValue` for integer data. In older versions of numpy (I think newer versions have fixed this) if you do `.where` with `np.nan` it would cast 32-bit floats to 64-bit floats. Now it knows that you really mean 32-bit float -> 32-bit float NaN. For integers, they would always be cast to 64-bit floats so the note in the documentation is trying to say "there is no reason for that so force 32-bit float conversion".

The note should probably be updated to refer to integer fields that are meant to stay integers like cloud type products or in this case 8-bit unsigned integer RGB images. If an RGB is being read in by Satpy and is provided as floating point data (instead of as an integer) then it should probably be normalized to a 0-1 range as that is the assumption by the trollimage package when saving images.

Reply:
Things like scale factor and add offset in a geotiff are not standard so I don't think we should add any support for that. Support for that should be added in a separate reader (a custom reader created by the user).

Geotiffs, at least those created by GDAL/rasterio, have `nodata` values set that should be masked. That should be considered the same as `_FillValue` in my opinion. In the future I'd like this reader to use the `rioxarray` library if possible to read GDAL datasets as xarray DataArrays.

As for not knowing if something is a category, assume that if it is an integer that there is a `nodata` value that can be used as `_FillValue`. If it integer and has an alpha band or has no `nodata` value specified, then don't set `_FillValue` and leave it as is.

In other readers, things like `units` are just metadata to fill in the space of missing information in the file. In those cases, the reader is specific to a particular implementation of a data file (VIIRS SDR in the HDF5 format versus generic HDF5 files) or are representing a standard (CF compliant NetCDF files). This generic image reader should try to use standard geotiff features.

Reply:
My original use for which I wrote this reader was to read any image as an image and use it as a background for `BackgroundCompositor` and to make it possible to resample them to a specific area. So in a way, I never considered it for anything else than categorical data.

Reply:
> Yeah, sure, but the question is what is the point of this pull request and how much work do you want to include in it? I am totally fine with the python code allowing for all sorts of flexibility as long as it is documented and that the defaults work. This is the "generic" image reader so none of this extra functionality should be required to get _something_ out of the reader. Adding options to a static list of datasets in the YAML sounds fine.
> 
> I think this reader should fix the dataset naming so it can be overridden in the YAML. If it includes some minimal nodata handling for geotiffs, then great. Anything beyond that should probably go in a separate PR, right?

My latest commit should address this points. There's now an option to control the handling of nodata. Default is to do nothing with the data but copying the nodata value (first element) to dataset.attrs['_FillValue']. 
Beside that you can add the option "nodata_handling" to the readers yaml dataset sections. If it is set to "nan_mask", the data is converted to float32 and masked with np.nan.

>  Anything beyond that should probably go in a separate PR, right?

Absolutely right. Anything else should be in another PR. But I need this together do get my stuff working and it's hard to handle multiple PRs addressing the same file. 

If you have any proposals for better naming of this option - you're welcome :-)
Hope this commit fulfills the demands.

Reply:
> My original use for which I wrote this reader was to read any image as an image and use it as a background for `BackgroundCompositor` and to make it possible to resample them to a specific area. So in a way, I never considered it for anything else than categorical data.

It's good to have such reader. Everything starts small. I hope that your needs are also covered with this modifications and the code still doesn't get too complex. These are only small changes and it may be more confusing to implement this in a separate reader.